_posts/2021-05-22-fluxv2-openfaas.md

```diff
@@ -0,0 +1,332 @@
+```sh 
+# Check if flux was able to install your manifests
+# If this shows an error then you need to fix it! its usually formatting issues
+kubectl get kustomization -n flux-system
+
+# Check if we could download the helm chart for openfaas
+# If this doesn't show "ready=True" then check the HelmRepository is correct
+kubectl get helmchart -n flux-system
+
+# To check the HelmRepository
+kubectl get helmrepository -n flux-system
+
+# Finally, check the HelmRelease, if this shows errors then you need to work out how to fix!
+kubectl describe helmrelease -n flux-system openfaas
+```
+
+# Deploy our first Function
+We are now going to generate a `Function` CRD which tells OpenFaaS to deploy a function, for speed we are going to use a
+pre-built store function as an example. You can use your existing functions or build new ones if you prefer.
+
+> to generate Function CRD files for your existing `stack.yaml` functions you can use `faas-cli generate --yaml stack.yaml`
+
+```sh 
+faas-cli generate --namespace openfaas-fn --from-store nodeinfo > nodeinfo.yaml
+faas-cli generate --namespace openfaas-fn --from-store figlet > figlet.yaml
+```
+
+Add these files to git and push 
+
+```sh 
+git add figlet.yaml nodeinfo.yaml
+git commit -m "Add openfaas functions"
+git push
+```
+
+You should see flux pull the latest git revision and install these CRDs.
+
+```sh 
+$ kubectl get functions
+NAME       AGE
+figlet     89s
+nodeinfo   89s
+
+$ kubectl get pods -n openfaas-fn
+NAME                        READY   STATUS    RESTARTS   AGE
+figlet-56475b7d8-8xwcq      1/1     Running   0          27s
+nodeinfo-6b54b47769-g7gkk   1/1     Running   0          27s
+```
+
+# Update our function
+To see the power of flux and GitOps in action we can delete the `functions` CRDs from our cluster and 
+watch as Flux puts them right back for us. 
+
+```sh 
+$ kubectl delete functions -n openfaas-fn --all
+function.openfaas.com "figlet" deleted
+function.openfaas.com "nodeinfo" deleted
+
+
+$ kubectl get functions -n openfaas-fn
+No resources found in openfaas-fn namespace.
+
+# Reconcile our flux kustomization, for speed, but this will happen
+# automatically every 10 minutes (which is configurable)
+
+$ flux reconcile kustomization flux-system
+► annotating Kustomization flux-system in flux-system namespace
+✔ Kustomization annotated
+◎ waiting for Kustomization reconciliation
+✔ Kustomization reconciliation completed
+✔ applied revision main/8dd6da7a8deaec21d3c7036f35caaaebf547177b
+
+$ kubectl get functions -n openfaas-fn
+NAME       AGE
+figlet     37s
+nodeinfo   37s
+```
+
+The functions were re-added by flux, it recognised that the objects that existed in Git didn't exist in the 
+cluster and added them back.
+
+As you can see, there is no more wondering if you applied the latest changes to a cluster, go and check Git
+and the Flux CRD objects to see what git revision was applied and if it succeeded.
+
```

Comment:
I feel like there is an element missing here where there is a helm chart for the functions.

Generating a helm chart for the functions means that the versions can be updated dynamically or as new versions of images are pushed to a registry.

I don't mind if you want to mention how to use kustomise, but unless I'm missing something, we are only deploying a static function manifest and not doing Continuous Deployment.

This is part of the value proposition of Flux v1/v2.